cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -424,19 +433,36 @@ withIndexEntries :: Index -> ([IndexCacheEntry] -> IO a) -> IO a
 withIndexEntries (RepoIndex repoCtxt repo@RepoSecure{..}) callback =
     repoContextWithSecureRepo repoCtxt repo $ \repoSecure ->
       Sec.withIndex repoSecure $ \Sec.IndexCallbacks{..} -> do
-        let mk :: (Sec.DirectoryEntry, fp, Maybe (Sec.Some Sec.IndexFile))
-               -> IO [IndexCacheEntry]
+        let mk :: (Sec.DirectoryEntry, fp, Maybe (Sec.Some Sec.IndexFile)) -> IO [IndexCacheEntry]
             mk (_, _fp, Nothing) =
               return [] -- skip unrecognized file
             mk (_, _fp, Just (Sec.Some (Sec.IndexPkgMetadata _pkgId))) =
               return [] -- skip metadata
-            mk (dirEntry, _fp, Just (Sec.Some (Sec.IndexPkgCabal pkgId))) = do
+            mk (dirEntry, _fp, Just (Sec.Some file@(Sec.IndexPkgCabal pkgId))) = do
               let blockNo = fromIntegral (Sec.directoryEntryBlockNo dirEntry)
-              return [CachePackageId pkgId blockNo]
+              timestamp <- Sec.indexEntryTime `fmap` indexLookupFileEntry dirEntry file
+              return [CachePackageId pkgId blockNo timestamp]
             mk (dirEntry, _fp, Just (Sec.Some file@(Sec.IndexPkgPrefs _pkgName))) = do
+              let blockNo = fromIntegral (Sec.directoryEntryBlockNo dirEntry)
               content <- Sec.indexEntryContent `fmap` indexLookupFileEntry dirEntry file
-              return $ map CachePreference (parsePreferredVersions content)
-        entriess <- lazySequence $ map mk (Sec.directoryEntries indexDirectory)
+              timestamp <- Sec.indexEntryTime `fmap` indexLookupFileEntry dirEntry file
+              return $ map (\x ->  CachePreference x blockNo timestamp) (parsePreferredVersions content)
+
+        -- TODO/FIXME: turn into lazy non-reversed list-generation
+        let go :: [(Sec.DirectoryEntry, Sec.IndexPath, Maybe (Sec.Some Sec.IndexFile))]
+                  -> Sec.DirectoryEntry
+                  -> IO [(Sec.DirectoryEntry, Sec.IndexPath, Maybe (Sec.Some Sec.IndexFile))]
+            go acc dent = do
+                (sie0, mde) <- indexLookupEntry dent
+                let e = case sie0 of Sec.Some sie -> (dent, Sec.indexEntryPath sie, fmap Sec.Some (Sec.indexEntryPathParsed sie))
+                    acc' = e:acc
+                maybe (return acc') (go acc') mde
+
+        -- entriess <- lazySequence $ map mk (Sec.directoryEntries indexDirectory)
+        -- we need all historic entries, including the overwritten ones
+        rents <- go [] (Sec.directoryFirst indexDirectory)
```

Comment:
I don't understand this bit. What was wrong with using `directoryEntries`? It looks like we're just re-implenting something equivalent locally? What am I missing?